src/sysext/sysext.c

```diff
@@ -273,6 +273,106 @@ static int need_reload(
+        if (n_submounts == 0)
+                return 0;
+
+        _cleanup_free_ char *dir_stack = path_join(stack, "dir");
+        if (!dir_stack)
+                return log_oom();
+
+        r = mkdir_p(dir_stack, 0700);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create directory %s: %m", dir_stack);
+
+        _cleanup_free_ char *file_stack = path_join(stack, "file");
+        if (!file_stack)
+                return log_oom();
+
+        r = mkdir_parents(file_stack, 0700);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create parent directories of %s: %m", file_stack);
+
+        r = touch(file_stack);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create %s: %m", file_stack);
+
+        FOREACH_ARRAY(m, submounts, n_submounts) {
+                _cleanup_free_ char *t = NULL;
+                const char *suffix;
+                struct stat st;
+
+                assert_se(suffix = path_startswith(m->path, src));
+
+                t = path_join(dst, suffix);
+                if (!t)
+                        return -ENOMEM;
+
+                if (fstat(m->mount_fd, &st) < 0)
+                        return log_error_errno(errno, "Failed to stat %s: %m", m->path);
+
+                r = mkdir_parents(t, 0755);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create parent directories of %s: %m", t);
+
+                if (S_ISDIR(st.st_mode)) {
+                        if (mkdir(t, 0755) < 0 && errno != EEXIST)
+                                return log_error_errno(errno, "Failed to create directory %s: %m", t);
+                } else {
+                        r = touch(t);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to create file %s: %m", t);
+                }
+
+                /* Some submounts might be a stack of multiple mounts on top of the same path. We have to
+                 * make sure we move all of them to the destination. Since the kernel does not provide us
+                 * with an interface to move a stack mounts, we do it manually by popping the stack one by
+                 * one onto a temporary location (where we'll get the stack in reverse order), before popping
+                 * it again one by one to the desired location (which will be in the original order again).
+                 * Because we don't know upfront whether a mount will be stacked or not, we don't try to
+                 * optimize this for non-stacked mounts.
+                 */
+
+                stack = S_ISDIR(st.st_mode) ? dir_stack : file_stack;
+
+                while (path_is_mount_point(m->path)) {
+                        r = mount_follow_verbose(LOG_DEBUG, m->path, stack, NULL, MS_BIND|MS_REC, NULL);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to mount %s to %s: %m", m->path, stack);
```

Comment:
the mount_xyz_verbose() calls will already generate a proper log message for you on failure, just specify LOG_ERR as first arg, and then change the return to a simple `return r`

Reply:
hmm, if you bind mount this is MS_REC, then this means that any contained child mounts will be bind over already, but you are not taking this into consideration afaics and will bind mount them manually a 2nd time?

Reply:
`get_sub_mounts()` doesn't recurse, it's meant to be used with `MS_REC`